[HSEARCH] 7.1.0.CR1 Release (#189)
hibernate/in.relation.to: posts/MarkoB/2024-02-12-hibernate-search-7-1-0-CR1.adoc

= Hibernate Search 7.1.0.CR1 is out
Marko Bekhta
:awestruct-tags: [ "Hibernate Search", "Lucene", "Elasticsearch", "Releases" ]
:awestruct-layout: blog-post
:hsearch-doc-url-prefix: https://docs.jboss.org/hibernate/search/7.1/reference/en-US/html_single/
:hsearch-getting-started-orm-url-prefix: https://docs.jboss.org/hibernate/search/7.1/getting-started/orm/en-US/html_single/
:hsearch-getting-started-stanadlone-url-prefix: https://docs.jboss.org/hibernate/search/7.1/getting-started/standalone/en-US/html_single/
:hsearch-jira-url-prefix: https://hibernate.atlassian.net/browse
:hsearch-version-family: 7.1
:hsearch-jira-project-id: 10061
:hsearch-jira-version-id: 32251
---

We just published Hibernate Search 7.1.0.CR1,
a first candidate release of the next minor version of Hibernate Search.

This version brings a lot of updates and improvements for working with a Standalone POJO mapper,
new query string predicate,

> as well as takes the advantage of the new Elasticsearch version and removes some of the limitations on vector search capabilities.

Reviewer: ```suggestion
and takes advantage of Elasticsearch 8.12 to remove some of the limitations on vector search capabilities.
```

== What's new

[NOTE]
====
For a summary of all new features and improvements since 7.0,
head to https://hibernate.org/search/releases/7.1/#whats-new[the dedicated page on hibernate.org].
====

=== Dependency upgrades

[[orm-version]]
Hibernate ORM (link:{hsearch-jira-url-prefix}/HSEARCH-5078[HSEARCH-5078])::
Hibernate Search now depends on Hibernate ORM 6.4.4.Final.
[[lucene-version]]
Lucene (link:{hsearch-jira-url-prefix}/HSEARCH-5069[HSEARCH-5069])::

> The Lucene backend now uses Lucene 9.9.2.
> Which brings some bug fixes related to vector search.

Reviewer: ```suggestion
The Lucene backend now uses Lucene 9.9.2,
which brings some bug fixes related to vector search.
```

from type names. The type name for the standalone POJO mapper is the name used when registering an entity,
be it via the link:{hsearch-doc-url-prefix}#mapping-entitydefinition-name[`@SearchEntity(name=...)` annotation]
or through the link:{hsearch-doc-url-prefix}#mapping-entitydefinition-programmatic[programmatic definition] using `.searchEntity()`.

====
[source, Java, indent=0, subs="+attributes"]
----
SearchMapping searchMapping = /* ... */ // <1>
SearchScope<Book> bookScope = searchMapping.scope( Book.class ); // <2>
SearchScope<Person> personSubTypesScope = searchMapping.scope( Person.class,
        List.of( "Manager", "Associate" ) ); // <3>
----
<1> link:{hsearch-doc-url-prefix}#entrypoints-search-mapping[Retrieve the `SearchMapping`].
<2> Create a `SearchScope` targeting only the `Book` entity type using a class.
<3> Create a `SearchScope` targeting only the `Manager` and `Associate` subtypes using their entity names.
====

See this section of the reference documentation on the link:{hsearch-doc-url-prefix}#entrypoints-search-scope[ways to create search scope]

> for more information.

Reviewer: Not sure this is worth mentioning... maybe move it last?

// ...
    .build(); // <3>
----
<1> Create a builder, passing an `AnnotatedTypeSource` to let Hibernate Search know where to look for annotations.
+
Thanks to link:{hsearch-doc-url-prefix}#mapping-classpath-scanning[classpath scanning],
your `AnnotatedTypeSource` only needs to include one class
from each JAR containing annotated types.
Other types should be automatically discovered.
<2> Set additional configuration properties.
<3> Build the `SearchMapping`.
====

See this section of the reference documentation on the link:{hsearch-doc-url-prefix}#mapping-entitydefinition[entity definition]
for more information.

[[search-dsl-predicate-knn]]
=== `knn` predicate updates

With the 8.12 release of Elasticsearch, a new link:https://www.elastic.co/guide/en/elasticsearch/reference/8.12/query-dsl-knn-query.html[`knn` query]
was introduced. This query is more aligned with the vector search capabilities provided by Hibernate Search,
and integration with Elasticsearch was updated to use this new query.
This means that previous limitations imposed on the vector search when using the Elastic distribution of the Elasticsearch backend

> are now removed. It also means that the vector search on this distribution will only be available starting with the 8.12.0 version.

Reviewer: ```suggestion
are now removed.

It also means that vector search on Elasticsearch is now only available with Elasticsearch 8.12 and later; Hibernate Search no longer supports vector search on Elasticsearch 8.11 and earlier.
```

link:{hsearch-doc-url-prefix}#search-dsl-predicate-simple-query-string[`simpleQueryString` predicate].

====
[source, JAVA, indent=0, subs="+callouts"]
----
List<Book> hits = searchSession.search( Book.class )
    .where( f -> f.queryString().field( "description" )
        .matching( "robots +(crime investigation disappearance)^10 +\"investigation help\"~2 -/(dis)?a[p]+ea?ance/" ) )
    .fetchHits( 20 );
----
The query string, in this predicate, will result in a boolean query with 4 clauses:

* a should clause matching `robots`;
* two must clauses
** another boolean query constructed from `(crime || investigation || disappearance)` string with a boost of `10`
** a query matching the phrase `investigation help` with the phrase slop equals to `2`
* a must not clause matching a regular expression `(dis)?a[p]+ea?ance`

+
Note that each of the mentioned clauses may itself end up being translated into other types of queries.
====

See this section of the reference documentation on the link:{hsearch-doc-url-prefix}#search-dsl-predicate-query-string[`queryString` predicate]

> for more information.

Reviewer: These are the most important points IMO, I'd move them first.

[[other-changes]]
=== Other improvements and bug fixes

* link:{hsearch-jira-url-prefix}/HSEARCH-5020[HSEARCH-5020]:
Hibernate Search increases the maximum allowed vector dimension for a Lucene backend to `4096` to allow indexing longer vectors produced by some models.
* link:{hsearch-jira-url-prefix}/HSEARCH-5073[HSEARCH-5073]:

> Hibernate Search opens up an ability to specify link:{hsearch-doc-url-prefix}#search-dsl-predicate-simple-query-string-minimumshouldmatch[minimum-should-match parameter] on a simple query predicate.

Reviewer: ```suggestion
Hibernate Search now allows specifying a link:{hsearch-doc-url-prefix}#search-dsl-predicate-simple-query-string-minimumshouldmatch[minimum-should-match parameter] on a simple query predicate.
```
